Composer/plugins/azurePublish/src/node/azureProvisionUtil.ts

```diff
@@ -0,0 +1,323 @@
+        }
+        catch (err) {
+            this.logger({
+                status: AzureAPIStatus.ERROR,
+                message: JSON.stringify(err, Object.getOwnPropertyNames(err)),
+            });
+            return []
+        }
+    }
+
+    /**
+     * Get all resource groups under a subscription
+     */
+    public async GetResourceGroupBySubscription(subscriptionId: string): Promise<Array<ResourceGroup>> {
+        try {
+            if (subscriptionId) {
+                this.logger({
+                    status: AzureAPIStatus.PARAM_ERROR,
+                    message: 'Need subscription or subscription id as a parameter.'
+                });
+                return []
+            }
+            const resourceManagementClient = new ResourceManagementClient(this.credentials, subscriptionId);
+            const resourceGroupsResult = await resourceManagementClient.resourceGroups.list();
+            if (resourceGroupsResult._response.status >= 300) {
+                this.logger({
+                    status: AzureAPIStatus.ERROR,
+                    message: resourceGroupsResult._response.bodyAsText
+                });
+                return []
+            }
+            return resourceGroupsResult._response.parsedBody;
+        }
+        catch (err) {
+            this.logger({
+                status: AzureAPIStatus.ERROR,
+                message: JSON.stringify(err, Object.getOwnPropertyNames(err)),
+            });
+            return []
+        }
+    }
+
+    /**
+     * Get all resource groups under a subscription id
+     */
+    public async GetResourceGroupBySubscriptionId(subscriptionId: string): Promise<Array<ResourceGroup>> {
+        try {
+            if (!subscriptionId) {
+                this.logger({
+                    status: AzureAPIStatus.PARAM_ERROR,
+                    message: 'Need subscription or subscription id as a parameter.'
+                });
+                return []
+            }
+            const resourceManagementClient = new ResourceManagementClient(this.credentials, subscriptionId);
+            const resourceGroupsResult = await resourceManagementClient.resourceGroups.list();
+            if (resourceGroupsResult._response.status >= 300) {
+                this.logger({
+                    status: AzureAPIStatus.ERROR,
+                    message: resourceGroupsResult._response.bodyAsText
+                });
+                return []
+            }
+            return resourceGroupsResult._response.parsedBody;
+        }
+        catch (err) {
+            this.logger({
+                status: AzureAPIStatus.ERROR,
+                message: JSON.stringify(err, Object.getOwnPropertyNames(err)),
+            });
+            return []
+        }
+    }
+
+    /**
+     * Check bot channel registration name availability
+     */
+    public async CheckBotNameAvailability(botName: string, subscriptionId: string): Promise<CheckNameAvailabilityResponseBody> {
```

Comment:
If this is going to be used by the client, why is it here instead of along side the other /azure/ APIS in the provision controller?

Reply:
> If this is going to be used by the client, why is it here instead of along side the other /azure/ APIS in the provision controller?

I will move this to the controller folder, functions with api route could call this util class making api call to external api.